README.rst
fluentcms-jumbotron
===================

.. image:: https://github.com/django-fluent/fluentcms-jumbotron/actions/workflows/tests.yaml/badge.svg?branch=master
    :target: https://github.com/django-fluent/fluentcms-jumbotron/actions/workflows/tests.yaml
.. image:: https://img.shields.io/pypi/v/fluentcms-jumbotron.svg
    :target: https://pypi.python.org/pypi/fluentcms-jumbotron/
.. image:: https://img.shields.io/pypi/l/fluentcms-jumbotron.svg
    :target: https://pypi.python.org/pypi/fluentcms-jumbotron/
.. image:: https://img.shields.io/codecov/c/github/django-fluent/fluentcms-jumbotron/master.svg
    :target: https://codecov.io/github/django-fluent/fluentcms-jumbotron?branch=master

Displaying a Bootstrap 3 Jumbotron_ in a page


Installation
============

First install the module, preferably in a virtual environment. It can be installed from PyPI:

.. code-block:: bash

    pip install fluentcms-jumbotron

First make sure the project is configured for django-fluent-contents_.

Then add the following settings:

.. code-block:: python

    INSTALLED_APPS += (
        'fluentcms_jumbotron',
    )

    FLUENT_CONTENTS_PLACEHOLDER_CONFIG = {
        'slot name': {
            'plugins': ('JumbotronPlugin', ...),
        },
    }

The database tables can be created afterwards:

.. code-block:: bash

    ./manage.py migrate


Frontend styling
================

The jumbotron is rendered with the HTML that Bootstrap prescribes:

.. code-block:: html+django

    <div class="jumbotron">
      <h1>Hello, world!</h1>
      <p>...</p>
      <p><a class="btn btn-primary btn-lg" href="#" role="button">Learn more</a></p>
    </div>

The standard Bootstrap 3 CSS will provide a reasonable styling for this,
which can either be overwritten, or replaced in your own CSS files.
The defaults provided by Bootstap 3 is: https://github.com/twbs/bootstrap-sass/blob/master/assets/stylesheets/bootstrap/_jumbotron.scss

Customizing
-----------

Centering, adding backgrounds, etc.. all happen by adding CSS styling. For example:

.. code-block:: css

    .jumbotron {
      background: url('/static/frontend/images/background.jpg') no-repeat fixed 0 0;
      background-size: cover;
      color: #fff;
      text-align: center;
    }

    .jumbotron .btn {
      margin-top: 12px;  /* For Sass: $padding-base-vertical * 2; */
    }

When you use Sass, you can also override the Sass variables.

Full page width
---------------

To display the Bootstrap Jumbotron full page, you likely need to break out of the container
the ``JumbotronPlugin`` is rendered in. For example, when your page looks like:

.. code-block:: html+django

    <div class="container">
        {% page_placeholder "homepage" title="Homepage" role="m" %}
    </div>

You can change that into:

.. code-block:: html+django

    <div class="container">
        {% page_placeholder "homepage" title="Homepage" role="m" template="pages/placeholders/homepage.html" cachable=1 %}
    </div>

The ``pages/placeholders/homepage.html`` template looks like:

.. code-block:: html+django

    {% for contentitem, html in contentitems %}
      {% if contentitem.plugin.name == 'JumbotronPlugin' %}
        </div>
        {{ html }}
        <div class="container">
      {% else %}
        {{ html }}
      {% endif %}
    {% endfor %}

Note the exact HTML tags depend on your frontend HTML layout.

The ``cachable=1`` flag is a promise that the template always returns the same result for every request.
Otherwise, remove it.

Contributing
------------

If you like this module, forked it, or would like to improve it, please let us know!
Pull requests are welcome too. :-)

.. _django-fluent-contents: https://github.com/django-fluent/django-fluent-contents
.. _jumbotron: http://getbootstrap.com/components/#jumbotron


CHANGES.rst
Changelog
=========

Version 3.0 (2021-11-16)
------------------------

* Added Django 3 and 4 support
* Dropped Python 2.7 support.
* Dropped Django 1.10 / 1.11 support.


Version 2.0 (2018-01-22)
------------------------

* Added Django 2.0 support.
* Dropped Django 1.7 / 1.8 / 1.9 support


Version 1.1.1 (2016-08-19)
--------------------------

* Fix Django migration changes for ``verbose_name`` fields in migrations.


Version 1.1 (2016-06-19)
------------------------

* Added second button.


Version 1.0 (2016-05-15)
------------------------

* First version


fluentcms_jumbotron/tests/models.py
from django.db import models
from fluent_contents.models import PlaceholderField


class ExamplePage(models.Model):
    slug = models.SlugField()
    content = PlaceholderField(slot="content")

    def __str__(self):
        return self.slug


fluentcms_jumbotron/tests/test_page.py
from django.template import Context, Template
from django.test import RequestFactory, TestCase, override_settings
from django.utils.html import strip_tags
from django.utils.text import slugify
from fluent_contents.models import Placeholder

from fluentcms_jumbotron.models import JumbotronItem

from .models import ExamplePage


class JumbotronTests(TestCase):
    """
    Testing jumbotrons
    """

    @classmethod
    def create_page(cls, slug, **jumbotron_kwargs):
        """
        Factory to create an page
        """
        page = ExamplePage.objects.create(slug=slug)
        placeholder = Placeholder.objects.create_for_object(page, slot="content")
        JumbotronItem.objects.create_for_placeholder(placeholder, **jumbotron_kwargs)
        return page

    @override_settings(FLUENT_CONTENTS_CACHE_OUTPUT=False)  # in runtests.py disabled
    def test_default_jumbotron(self):
        """
        Default jumbotron rendering
        """
        page1 = self.create_page(
            slug="page1",
            title="Hello",
            content="<p>test</p>",
            button1_url="http://example.org/",
            button1_title="GO",
        )

        template = Template('{% load fluent_contents_tags %}{% page_placeholder "content" %}')
        request = RequestFactory().get("/", HTTP_HOST="example.org")

        html = template.render(Context({"page": page1, "request": request}))
        expected = """<div class="jumbotron">
  <div class="container">
    <h1>Hello</h1>
    <p>test</p>
    <p><a class="btn btn-primary btn-lg" href="http://example.org/" role="button">GO</a></p>
  </div>
</div>"""
        self.assertEqual(slugify(strip_tags(html)), "hello-test-go")
        self.assertHTMLEqual(html.strip(), expected)


fluentcms_jumbotron/tests/migrations/0001_initial.py
# Generated by Django 1.9.4 on 2016-05-09 14:56

from django.db import migrations, models


class Migration(migrations.Migration):

    initial = True

    dependencies = []

    operations = [
        migrations.CreateModel(
            name="ExamplePage",
            fields=[
                (
                    "id",
                    models.AutoField(
                        auto_created=True, primary_key=True, serialize=False, verbose_name="ID"
                    ),
                ),
                ("slug", models.SlugField()),
            ],
        ),
    ]


runtests.py
#!/usr/bin/env python
import sys

import django
from django.conf import global_settings as default_settings
from django.conf import settings
from django.core.management import execute_from_command_line

if not settings.configured:
    settings.configure(
        DATABASES={
            "default": {
                "ENGINE": "django.db.backends.sqlite3",
                "NAME": ":memory:",
            },
        },
        SITE_ID=1,
        INSTALLED_APPS=(
            "django.contrib.contenttypes",
            "django.contrib.auth",
            "django.contrib.sites",
            "fluent_contents",
            "fluentcms_jumbotron",
            "fluentcms_jumbotron.tests",
        ),
        MIDDLEWARE=(),
        TEMPLATES=[
            {
                "BACKEND": "django.template.backends.django.DjangoTemplates",
                "DIRS": (),
                "OPTIONS": {
                    "loaders": (
                        "django.template.loaders.filesystem.Loader",
                        "django.template.loaders.app_directories.Loader",
                    ),
                    "context_processors": (
                        "django.template.context_processors.debug",
                        "django.template.context_processors.i18n",
                        "django.template.context_processors.media",
                        "django.template.context_processors.request",
                        "django.template.context_processors.static",
                        "django.contrib.auth.context_processors.auth",
                    ),
                },
            },
        ],
        FLUENT_CONTENTS_CACHE_OUTPUT=False,
        TEST_RUNNER="django.test.runner.DiscoverRunner",
        DEFAULT_AUTO_FIELD="django.db.models.BigAutoField",
    )

DEFAULT_TEST_APPS = [
    "fluentcms_jumbotron",
]


def runtests():
    other_args = list(filter(lambda arg: arg.startswith("-"), sys.argv[1:]))
    test_apps = (
        list(filter(lambda arg: not arg.startswith("-"), sys.argv[1:])) or DEFAULT_TEST_APPS
    )
    argv = sys.argv[:1] + ["test", "--traceback"] + other_args + test_apps
    execute_from_command_line(argv)


if __name__ == "__main__":
    runtests()


fluentcms_jumbotron/models.py
from django.conf import settings
from django.db import models
from django.utils.functional import cached_property
from django.utils.translation import gettext_lazy as _
from fluent_contents.extensions import PluginHtmlField, PluginUrlField
from fluent_contents.models import ContentItem


class JumbotronItem(ContentItem):
    """
    Pager item, to show a previous/next page.
    The pages are auto determined, but can be overwritten
    """

    title = models.CharField(_("Title"), max_length=200)
    content = PluginHtmlField(_("Content"))  # TODO: replace with child elements later?
    button1_title = models.CharField(_("Button 1 Title"), max_length=200, blank=True, null=True)
    button1_url = PluginUrlField(_("Button 1 URL"), blank=True, null=True)

    button2_title = models.CharField(_("Button 2 Title"), max_length=200, blank=True, null=True)
    button2_url = PluginUrlField(_("Button 2 URL"), blank=True, null=True)

    class Meta:
        verbose_name = _("Jumbotron")
        verbose_name_plural = _("Jumbotron")

    def __str__(self):
        return self.title


fluentcms_jumbotron/content_plugins.py
"""
Definition of the plugin.
"""
from django.utils.translation import gettext_lazy as _
from fluent_contents.extensions import ContentPlugin, plugin_pool

from .models import JumbotronItem


@plugin_pool.register
class JumbotronPlugin(ContentPlugin):
    """
    CMS plugin for previous/next navigation element.
    """

    category = _("Landing page")
    model = JumbotronItem
    render_template = "fluentcms_jumbotron/jumbotron.html"
    fieldsets = (
        (
            None,
            {
                "fields": ("title", "content"),
            },
        ),
        (
            _("Buttons"),
            {
                "fields": (
                    ("button1_title", "button2_title"),
                    ("button1_url", "button2_url"),
                ),
            },
        ),
    )


fluentcms_jumbotron/migrations/0002_add_button2.py
import fluent_contents.extensions
from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ("fluentcms_jumbotron", "0001_initial"),
    ]

    operations = [
        migrations.RenameField(
            model_name="jumbotronitem",
            old_name="button_title",
            new_name="button1_title",
        ),
        migrations.RenameField(
            model_name="jumbotronitem",
            old_name="button_url",
            new_name="button1_url",
        ),
        migrations.AddField(
            model_name="jumbotronitem",
            name="button2_title",
            field=models.CharField(
                max_length=200, null=True, verbose_name="Button 2 Title", blank=True
            ),
        ),
        migrations.AddField(
            model_name="jumbotronitem",
            name="button2_url",
            field=fluent_contents.extensions.PluginUrlField(
                max_length=300, null=True, verbose_name="Button 2 URL", blank=True
            ),
        ),
    ]


fluentcms_jumbotron/migrations/0001_initial.py
import fluent_contents.extensions
from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ("fluent_contents", "0001_initial"),
    ]

    operations = [
        migrations.CreateModel(
            name="JumbotronItem",
            fields=[
                (
                    "contentitem_ptr",
                    models.OneToOneField(
                        parent_link=True,
                        auto_created=True,
                        primary_key=True,
                        serialize=False,
                        to="fluent_contents.ContentItem",
                        on_delete=models.CASCADE,
                    ),
                ),
                ("title", models.CharField(max_length=200, verbose_name="Title")),
                ("content", fluent_contents.extensions.PluginHtmlField(verbose_name="Content")),
                (
                    "button_title",
                    models.CharField(
                        max_length=200, null=True, verbose_name="Button 1 Title", blank=True
                    ),
                ),
                (
                    "button_url",
                    fluent_contents.extensions.PluginUrlField(
                        max_length=300, null=True, verbose_name="Button 1 URL", blank=True
                    ),
                ),
            ],
            options={
                "db_table": "contentitem_fluentcms_jumbotron_jumbotronitem",
                "verbose_name": "Jumbotron",
                "verbose_name_plural": "Jumbotron",
            },
            bases=("fluent_contents.contentitem",),
        ),
    ]
